### DISCUSSION TITLE
twoaxistracking - a new package in the pvlib organization?

### DISCUSSION THREAD

**Discussion by AdamRJensen** [pvlib/pvlib-python#1407] [General]
It seems that simulating shading of two-axis trackers is more difficult than expected, and since I never got a reply from my post in the Google group I decided to develop a package. The package is called [twoaxistracking](https://github.com/AdamRJensen/twoaxistracking) and is documented at [twoaxistracking.readthedocs.io](https://twoaxistracking.readthedocs.io/).

The repository is currently under my GH user, which is less than ideal for something that hopefully will get used by many others. Therefore, I was wondering if there is interest in moving the package to the pvlib organization? After all, it seems to fit the scope, in that it deals with (1) solar/PV and (2) simulation.

I suppose this mostly depends on the aim of the pvlib organization is. Is it to become similar to organizations such as [sunpy](https://github.com/sunpy) which has many projects (I guess of which pvanalytics would be one)? Or is it solely dedicated to pvlib? 

I'm willing to be the maintainer long-term and with help from @kanderso-nrel the package will soon have a range of tests and linting. Also, starting from version 0.2.0 the package will be available on PyPi.

**Comment by cwhanse** [comment]
@AdamRJensen as an organization, pvlib is open to several projects that provide open-source software for solar power. pvlib/pvlib-python is the package for "forward" modeling, i.e., given weather data and system parameters, predict output. pvlib/pvanalytics is intended to be the package supporting "backward" analysis, i.e., providing functions that operate on measured data (weather, PV system output) to provide quality metrics, detect features such as inverter clipping, etc. There's hope to launch a third project pvlib/pvmodules to contain code to generate the CEC module parameter file.

Nothing in pvlib-python currently is specific to two-axis tracking, mostly because I think the use cases are not common. I am open to a new module in pvlib-python, adding code to pvlib/shading.py, or a combination of these options.

**Reply by AdamRJensen** [reply]
@cwhanse Thanks for your very informative answer!

I think it is wisest to have it be its own package, because (1) the code heavily relies on `shapely` and we don't want to introduce this dependency to pvlib-python, and (2) the use cases are relatively infrequent and perhaps does not justify the added complexity of including it in pvlib-python.

I'd be interested in the opinion of the other maintainers (perhaps just a thumbs up if you support the idea) and what the next steps would be if we decide to include it in the pvlib organization?

**Reply by kandersolar** [reply]
I'm less convinced than I used to be that shapely is a blocker for integrating with the pvlib repo (not org) -- it's also a dependency for `pvfactors`, so it's in `pvlib[optional]` regardless of what we do here. 

However I still lean towards a separate repository because of the auxiliary code for array layout and plotting.  To me the `twoaxistracking` repo seems closer to something like PVMismatch, which is sort of a standalone application, than it does a module of pvlib.  But I don't really feel strongly either way and would be happy to help integrate it with pvlib itself if that's where the group opinion is.

**Reply by mikofski** [reply]
I am also of the opinion that this is better as a separate package, regardless of the shapely dependency.

One really neat solution is to make it an optional subpackage. I think the way this works is...
1. it's in it's own repository
2. the name of the package is like `pvlib.twoaxis`
3. you can list it as an optional dependency in the pvlib `setup.py`
4. when it installs it will install just like it was part of pvlib, even though it's actually a separate package.

I only found a few references on how to create an optional sub-package or "namespace package distribution":
* [Packaging namespace packages](https://packaging.python.org/en/latest/guides/packaging-namespace-packages/)
* [Python `setup.py` define optional subpackages with optional dependencies](https://stackoverflow.com/questions/51245648/python-setup-py-define-optional-subpackages-with-optional-dependencies)

**Reply by kandersolar** [reply]
That subpackage concept is interesting. I wonder if it might be a good fit for `pvlib.iotools` too.

Anyway, shall we move the repository now and figure out the integration with pvlib later?  Haven't heard from @wholmgren here yet.

**Reply by wholmgren** [reply]
> shall we move the repository now and figure out the integration with pvlib later?

+1

**Reply by mikofski** [reply]
Do you want to (A) transfer ownership to pvlib? That will make pr’s naturally point to its new location & and keep issue/PR history versus (B-1) forking it or (B-2) cloning and creating a new repo which would leave the old location as “upstream” & which would lose all the old history.

My preference is A

**Reply by AdamRJensen** [reply]
And we are live! https://github.com/pvlib/twoaxistracking

readthedocs has been updated so it now points to the new location, and @kanderso-nrel has been added as co-maintainer.

**Reply by kandersolar** [reply]
Documenting a quirk of the process for posterity: @AdamRJensen couldn't transfer it directly to pvlib because of lack of permission to create new repositories in the pvlib org, so instead we did it in two steps using my account as a middleman.  🤷
